lib/llvm-ast.js

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2021, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export class LlvmAstParser {
+    constructor(compilerProps) {
+        this.maxAstLines = 500;
+        if (compilerProps) {
+            this.maxAstLines = compilerProps('maxLinesOfAst', this.maxAstLines);
+        }
+
+        this.spanTypes = {
+            NONE: 'none',
+            EMPTY: 'empty',
+            LEFT: 'left',
+            RIGHT: 'right',
+            FULL: 'full',
+        };
+    }
+
+    parseSpan(line) {
+        const spanRegex = /<((?:line|col)[\d ,:ceilno]+)>/;
+        const twoLineRegex = /line:(\d+)[,:].*line:(\d+)[:>]/;
+        const leftLineRegex = /^line:(\d+)/;
+        const rightLineRegex = / line:(\d+):/;
+
+        let m = line.match(spanRegex);
+        if (m) {
+            let span = m[1];
+            let ll = span.match(twoLineRegex);
+            if (ll) {
```

Comment:
Offtopic... wouldn't it be lovely if js had `if (const ll = ...; ll) {}` here? Oh well :(

Reply:
True. Such an omission for such a flexible language.